android/src/main/java/com/reactnativeteadssdkmodule/RNInReadAdPlacement.kt

```diff
@@ -0,0 +1,198 @@
+import tv.teads.sdk.renderer.InReadAdView
+import java.util.*
+
+
+class RNInReadAdPlacement(reactContext: ReactApplicationContext) :
+  ReactContextBaseJavaModule(reactContext) {
+
+  override fun getName(): String {
+    return "RNInReadAdPlacement"
+  }
+
+  private fun sendEvent(reactContext: ReactContext, eventName: String, params: WritableMap?) {
+    reactContext
+      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
+      .emit(eventName, params)
+  }
+
+  //requestAd
+  @ReactMethod
+  fun requestAd(pid: Int, settingsMap: ReadableMap, promise: Promise) {
+    val instanceIdentifier = UUID.randomUUID().toString()
+    RNAdInstanceManager.instance(pid).inReadAdPlacement.requestAd(AdRequestSettings.fromMap(
+      settingsMap.toHashMap().toMap()
+    ),
+      object : InReadAdListener {
+        override fun adOpportunityTrackerView(trackerView: AdOpportunityTrackerView) {
+          val params = Arguments.createMap()
+          params.putString("didPlay", "didPlay")
+          sendEvent(reactApplicationContext, "didPlay", params)
+        }
+
+        override fun onAdReceived(
+          inReadAdView: InReadAdView,
+          adRatio: AdRatio
+        ) {
+          RNAdInstanceManager.new(
+            RNAdInstanceManager.AdInstanceMap(
+              inReadAdView,
+              instanceIdentifier
+            )
+          )
+          promise.resolve(instanceIdentifier)
+          val params = Arguments.createMap()
+          params.putString("didReceiveAd", "didReceiveAd")
+          sendEvent(reactApplicationContext, "didReceiveAd", params)
+          Handler(Looper.getMainLooper()).post {
+            Log.d("from ad", "didReceiveAd")
+
+          }
+        }
+
+        override fun onAdClicked() {
+          val params = Arguments.createMap()
+          params.putString("didRecordClick", "didRecordClick")
+          sendEvent(reactApplicationContext, "didRecordClick", params)
+          Handler(Looper.getMainLooper()).post {
+            Log.d("form ad", "didRecordClick")
+          }
+        }
+
+        override fun onAdClosed() {
+          val params = Arguments.createMap()
+          params.putString("didClose", "didClose")
+          sendEvent(reactApplicationContext, "didClose", params)
+          Handler(Looper.getMainLooper()).post {
+            Log.d(
+              "from ad",
+              "didClose"
+            )
+          }
+        }
+
+        override fun onAdError(code: Int, description: String) {
+          val params = Arguments.createMap()
+          params.putString("didCatchError", "didCatchError")
+          sendEvent(reactApplicationContext, "didCatchError", params)
+          Handler(Looper.getMainLooper()).post {
+            Log.d(
+              "from ad",
+              "didCatchError"
+            )
+            Log.d("from error", description)
+          }
+        }
+
+        override fun onAdImpression() {
+          val params = Arguments.createMap()
```

Comment:
Are you intentionally sending non existing params? Is there some purpose on it?

Reply:
What do you mean about empty params, normally there is just the adId, to know from which ad it comes from.
And then when the publisher listens to this event on React part he could implement anything, on the listener.